### PR TITLE
Some fixes for `device` on GPU

### PR DESCRIPTION
**Pull request by Linux-cpp-lisp**
## Description

Bureaucratic fixes for running on GPU:

- make `normalize2mom` respect current device
- make `Gate` keeps its activations as submodules so they `.to` with it
- add `_get_device`
- Make `spherical_harmonics` a wrapper for `SphericalHarmonics` 
- make `SphericalHarmonics` scriptable so that it doesn't trace and pick up tensors on the wrong device

## How Has This Been Tested?
e3nn test suite, our code's tests (which were failing before on GPU)

## Types of changes

- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds or improves functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation improvement (updates to user guides, docstrings, or developer docs)

## Checklist:

- [X] I have read the [**CONTRIBUTING**](https://github.com/e3nn/e3nn/blob/main/CONTRIBUTING.md) document.
- [X] My code follows the code style of this project.
- [X] I have updated the documentation (if relevant).
- [X] I have added tests that cover my changes (if relevant).
- [X] All new and existing tests passed.
- [X] I have updated the [Changelog](https://github.com/e3nn/e3nn/blob/main/ChangeLog.md).

### CODE REVIEW

File: e3nn/math/normalize_activation.py [hunk 1]
```diff
@@ -18,8 +18,17 @@ def moment(f, n, dtype=None, device=None):
...
     def __init__(self, f, dtype=None, device=None):
         super().__init__()
+
+        # Try to infer a device:
+        if device is None and isinstance(f, torch.nn.Module):
+            # Avoid circular import
+            from e3nn.util._argtools import _get_device
+            device = _get_device(f)
+
         cst = 1 / moment(f, 2, dtype=dtype, device=device) ** 0.5
```
Comment (mariogeiger) [inline e3nn/math/normalize_activation.py:32]: If I remember well, before `cst` was a python float.
Isn't it simpler if cst is a float, no need to convert any type and so on?

Comment (Linux-cpp-lisp) [reply e3nn/math/normalize_activation.py]: Switched back :+1:

File: e3nn/o3/cartesian_spherical_harmonics.py [hunk 2]
```diff
@@ -13,42 +13,94 @@
...
-        self.l = l
         self.normalize = normalize
         self.normalization = normalization
+        assert normalization in ['integral', 'component', 'norm']
+
+        irreps_in = Irreps(irreps_in)
+        if irreps_in not in (Irreps("1x1o"), Irreps("1x1e")):
+            raise ValueError(f"irreps_in for SphericalHarmonics must be either a vector (`1x1o`) or a psuedovector (`1x1e`), not `{irreps_in}`")
+        self.irreps_in = irreps_in
+        input_p = irreps_in[0][1][1]
```
Comment (mariogeiger) [inline e3nn/o3/cartesian_spherical_harmonics.py:45]: irreps_in[0].ir.p

File: e3nn/math/normalize_activation.py [hunk 3]
```diff
@@ -18,17 +18,28 @@ def moment(f, n, dtype=None, device=None):
...
-        cst = 1 / moment(f, 2, dtype=dtype, device=device) ** 0.5
+
+        # Try to infer a device:
+        if device is None and isinstance(f, torch.nn.Module):
+            # Avoid circular import
+            from e3nn.util._argtools import _get_device
+            device = _get_device(f)
+
+        with torch.no_grad():
+            cst = float(1 / moment(f, 2, dtype=dtype, device=device) ** 0.5)
```
Comment (mariogeiger) [inline e3nn/math/normalize_activation.py:34]:
```suggestion
            cst = moment(f, 2, dtype=dtype, device=device).pow(-0.5).item()
```

**Comment by Linux-cpp-lisp** [comment]
In general, it would probably be good if the `e3nn` tests ran on GPU when a GPU is available — this could catch some bugs sooner.

**Comment by mariogeiger** [comment]
Agreed

**Comment by Linux-cpp-lisp** [comment]
@mariogeiger filed an issue: https://github.com/e3nn/e3nn/issues/226
